File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/2472-maximum-number-of-non-overlapping-palindrome-substrings.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        ans = start = 0
        for r in range(k - 1, len(s)):
            # size k
            l = r - k + 1
            if l >= start and is_pal(l, r, s):
                ans += 1
                start = r + 1
                continue
            
            # size k + 1
            l = r - k
            if l >= start and is_pal(l, r, s):
                ans += 1
                start = r + 1
        
        return ans
        
def is_pal(l, r: int, s: str) -> bool:
    while l < r:
        if s[l] != s[r]:
            return False
        l += 1
        r -= 1
    
    return True

# greedy (the greedy idea is from editorial)
# time: O(nk)
# space: O(1)

File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/2472-maximum-number-of-non-overlapping-palindrome-substrings.go
func maxPalindromes(s string, k int) int {
    ans, start := 0, 0
    for right := k - 1; right < len(s); right++ {
        // palindrome with size k
        left := right - k + 1
        if left >= start && isPal(left, right, s) {
            ans++
            start = right + 1
            continue
        }

        // palindrome with size k + 1
        left = right - k
        if left >= start && isPal(left, right, s) {
            ans++
            start = right + 1
        }
    }

    return ans    
}

func isPal(l, r int, s string) bool {
    for l < r {
        if s[l] != s[r] {
            return false
        }
        l++
        r--
    }

    return true 
}

// greedy (the greedy idea is from editorial)
// time: O(nk)
// space: O(1)